vision/module/location.py

```diff
@@ -249,6 +251,171 @@ def _get_slopes(self) -> None:
+        lr_bound = (max_x, max_y)  # lower right corner
+
+        # Filter out rectangles that are not directly below text
+        circle_filter = np.array(
+            [], dtype=np.uint16
+        )  # array of indices of circles out of bounds
+
+        i = 0
+        for x, y, _ in self.circles:
+            if y <= lr_bound[1]:  # remove if above lower bound of text
+                circle_filter = np.append(circle_filter, i)
+            elif (
+                x <= ll_bound[0] or x >= lr_bound[0]
+            ):  # remove if horizontally outside text
+                circle_filter = np.append(circle_filter, i)
+            i += 1
+
+        self.circles = np.delete(self.circles, circle_filter, axis=0)
+
+    def _filter_distant_circles(self, depth_threshold: int) -> np.ndarray:
+        """
+        Filters circles that are too far away based on the depth image
+
+        Parameters
+        ----------
+        depth_threshold: int
+            The depth distance (in millimeters) at which to consider circles too far away
+
+        Returns
+        -------
+        None
+        """
+        # Create filter of far away circles based on depth image
+        circle_filter = np.array(
+            [], dtype=np.uint16
+        )  # array of indices of distant circles
+
+        i = 0
+        for x, y, _ in self.circles:
+            if (
+                self.depth[y][x] > depth_threshold
+            ):  # if too far away, remove (add index to filter)
+                circle_filter = np.append(circle_filter, i)
+            i += 1
+
+        self.circles = np.delete(self.circles, circle_filter, axis=0)
+
+    def _cluster_circles(self) -> None:
+        """
+        Groups circles that are near each other, and filters/removes ones that are and not near other circles
+
+        Returns
+        -------
+        None
+        """
+        MIN_CIRCLES = 5  # minimum number of total circles to require clusting
+        CLSTR_CRITERIA = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 10, 1.0)
+        COMPACT_THRESH = 40000  # NOTE: May need scaling up for higher res, only tested with 480p data
+        MAX_CLUSTERS = 30
+
+        if self.circles.shape[0] < 5:  # not enough circles for clustering
+            return self.circles
+
+        num_clusters = 2
+        compactness = COMPACT_THRESH + 1  # init compactness to enter while loop
+        while compactness > COMPACT_THRESH:  # ensure low variance within clusters
+            compactness, labels, _ = cv2.kmeans(
+                data=self.circles.astype(np.float32),  # kmeans requires float32
+                K=num_clusters,
+                bestLabels=None,
+                criteria=CLSTR_CRITERIA,
+                attempts=30,
+                flags=cv2.KMEANS_PP_CENTERS,
+            )
+            if num_clusters > MAX_CLUSTERS:
+                return self.circles
+            num_clusters += 1  # try higher K value
+        num_clusters -= 1  # decrement extra post-increment
```

Comment:
Can make this a bit more clear by replacing the increment/decrement with
```
if compactness > COMPACT_THRESH:
   num_clusters += 1
```
within the while loop